portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,87 @@
 
 package com.google.sps.servlets;
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.List;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+private List<String> comments;
+
+  @Override
+  public void init() {
+    comments = new ArrayList<>();
+  }
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String commentJson = convertToJson(comments);
     response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    response.getWriter().println(commentJson);
   }
-}
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    String comment = getCommentLeft(request);
+    //Warn about invalid comment
+    if (comment == null) {
+        // response.setContentType("text/html");
+        // response.getWriter().println("Please enter a valid comment");
+    } else {
+        comments.add(comment);
+        long timestamp = System.currentTimeMillis();
+
+        Entity taskEntity = new Entity("Comment");
+        taskEntity.setProperty("comment", comment);
+        taskEntity.setProperty("timestamp", timestamp);
+
+        datastore.put(taskEntity);
+
+        // Redirect back to the HTML page.
+        response.sendRedirect("/index.html");
+    }
+  }
+
+  /** Returns the choice entered by the player, or -1 if the choice was invalid. */
+  private String getCommentLeft(HttpServletRequest request) {
+    // Get the input from the form.
+    String commentLeftString = request.getParameter("comment-left");
+    
+    // Check for a valid input.
+    if (commentLeftString.equals("")) {
+        return null;
+    } else{
+       return commentLeftString; 
+    }
+  }
+
+  /**
+   * Converts a ServerStats instance into a JSON string using manual String concatentation.
+   */
+  private String convertToJson(List<String> arr) {
```

Comment:
I'd recommend just calling the "gson" library's version since you've already added it as a dependency in your pom.xml, that way you won't need to import the "List" library either.

```
import com.google.gson.Gson; // somewhere at the top of your file with your other imports
...
String json = gson.toJson(comments); // instead of convertToJson
```
(I know the tutorial said that you could convert Lists to json by putting together the string manually or using gson, but in reality, building json strings yourself is pretty tedious and hard to do without bugs)